grpc-client-server/prober_server/main_test.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func dialer() func(context.Context, string) (net.Conn, error) {
+	listener := bufconn.Listen(1024 * 1024)
+
+	server := grpc.NewServer()
+
+	pb.RegisterProberServer(server, &Server{})
+
+	go func() {
+		if err := server.Serve(listener); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	return func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+}
+
+func TestDoProbes(t *testing.T) {
+	tests := map[string]struct {
+		req       *pb.ProbeRequest
+		err       error
+		failedReq int32
+	}{
+		"simple": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "https://www.google.com",
+				NumberOfRequests: 1,
+			},
+			failedReq: 0,
+		},
+		"error": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "http://musterror:8080",
+				NumberOfRequests: 3,
+			},
+			failedReq: 3,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithTransportCredentials(insecure.NewCredentials()))
+			require.NoError(t, err)
+			defer conn.Close()
+
+			client := pb.NewProberClient(conn)
+
+			resp, err := client.DoProbes(context.Background(), tc.req)
+
+			if err == nil {
+				require.NotNil(t, resp.TtfbAverageResponseTime)
+			}
+
+			require.Equal(t, tc.failedReq, resp.FailedRequests)
+		})
+	}
+
+	t.Run("response time", func(t *testing.T) {
+		conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		require.NoError(t, err)
+		defer conn.Close()
+
+		client := pb.NewProberClient(conn)
+
+		timeSince = func(t time.Time) time.Duration {
```

Comment:
Rather than setting a global variable here, can you make it so that when you construct your server, you pass in this value? (Or actually, you pass in the `http.RoundTripper` implementation you want to use? Think about how the `http.Client` `struct` works - if you construct one with a custom `RoundTripper`, it'll use it, if not, it'll use a default one).

That way, in your test you can pass in a custom `RoundTripper` without needing any global variables.

Reply:
This is great, thanks! implemented now with `timeSince` mock for server and http client.

Not super happy with the implementation that client is referencing a variable from the parent function, will have to do some refactoring to remove that dependency and you suggesting of passing a custom transport to the server sounds like it might solve this issue..

```
	ttfbTotal := time.Duration(0)
	ttlbTotal := time.Duration(0)
	failed := 0
	latestTtfb := time.Duration(0)

	c := &http.Client{
		Transport: &TimedRoundTripper{
			underlying: http.DefaultTransport,
			recordTime: func(t time.Duration) {
				latestTtfb = t
				ttfbTotal += t
			},
			timeSince: s.timeSince,
		},
	}
```